File: examples/diffusion_elliptic.py
from __future__ import print_function
import numpy, os, sys
import scipy.sparse.linalg as sp_lin_alg

try:
  import HyperHDG
except (ImportError, ModuleNotFoundError) as error:
  sys.path.append(os.path.dirname(os.path.abspath(__file__)) + "/../import")
  import HyperHDG

# Importing done

poly_degree = 1
hyEdge_dim  = 1
cube_dim    = 2
refinement  = 1
debug_mode  = True
  
hdg_config                 = HyperHDG.config()
hdg_config.global_loop     = "Elliptic"
hdg_config.topology        = f'Cubic<{hyEdge_dim},{cube_dim}>'
hdg_config.geometry        = f'UnitCube<{hyEdge_dim},{cube_dim},double>'
hdg_config.node_descriptor = f'Cubic<{hyEdge_dim},{cube_dim}>'
hdg_config.local_solver    = "Diffusion<" + str(hyEdge_dim) + "," + str(poly_degree) + "," + str(
  2*poly_degree) + ",HG<" + str(hyEdge_dim) + ">::DiffusionElliptic,double>"
hdg_config.include_files   = ["examples/parameters/diffusion.hxx"]
hdg_config.cython_replacements = ["vector[unsigned int]", "vector[unsigned int]"]
hdg_config.debug_mode      = debug_mode

hyperHDGClass = HyperHDG.include(hdg_config)
HDGObject   = hyperHDGClass ( [2 ** refinement] * cube_dim )

vectorRHS = numpy.multiply( HDGObject.residual_flux(HDGObject.zero_vector()), -1. )

system_size = HDGObject.size_of_system()
A = sp_lin_alg.LinearOperator((system_size,system_size), matvec=HDGObject.trace_to_flux)

[vectorSolution, num_iter] = sp_lin_alg.cg(A, vectorRHS, tol=1e-13)
if num_iter != 0:
  print("CG solver failed with a total number of ", num_iter, "iterations.")
  raise RuntimeError("Linear solvers did not converge!")

print("Error: ", HDGObject.errors(vectorSolution)[0])

HDGObject.plot_option( "fileName" , "diffusion_elliptic_py" )
HDGObject.plot_option( "printFileNumber" , "false" )
HDGObject.plot_option( "scale" , "0.95" )
HDGObject.plot_solution(vectorSolution)

File: examples/diffusion_elliptic.cxx
#include <HyperHDG/geometry/unit_cube.hxx>
#include <HyperHDG/global_loop/elliptic.hxx>
#include <HyperHDG/local_solver/diffusion_ldgh.hxx>
#include <HyperHDG/node_descriptor/cubic.hxx>
#include <HyperHDG/sparse_la.hxx>
#include <examples/parameters/diffusion.hxx>

int main()
{
  const unsigned int poly_degree = 1;
  const unsigned int hyEdge_dim = 1;
  const unsigned int space_dim = 2;
  const unsigned int refinement = 1;

  GlobalLoop::Elliptic<Topology::Cubic<hyEdge_dim, space_dim>,
                       Geometry::UnitCube<hyEdge_dim, space_dim>,
                       NodeDescriptor::Cubic<hyEdge_dim, space_dim>,
                       LocalSolver::Diffusion<hyEdge_dim, poly_degree, 2 * poly_degree,
                                              HG<hyEdge_dim>::DiffusionElliptic, double> >
    HDG_wrapper(std::vector<unsigned int>(space_dim, 1 << refinement));

  std::vector<double> vectorRHS = HDG_wrapper.residual_flux(HDG_wrapper.zero_vector());
  for (unsigned int i = 0; i < vectorRHS.size(); ++i)
    vectorRHS[i] *= -1.;

  std::vector<double> vectorSolution = SparseLA::conjugate_gradient(vectorRHS, HDG_wrapper);

  std::cout << "Error: " << HDG_wrapper.errors(vectorSolution)[0] << std::endl;

  HDG_wrapper.plot_option("fileName", "diffusion_elliptic_c++");
  HDG_wrapper.plot_option("printFileNumber", "false");
  HDG_wrapper.plot_option("scale", "0.95");
  HDG_wrapper.plot_solution(vectorSolution);

  return 0;
}
